Fix jit can not recognize the imported function

qihqi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #63248

Summary:
Fixes https://github.com/pytorch/pytorch/issues/38964

Test Plan:
Unit test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30311185](https://our.internmc.facebook.com/intern/diff/D30311185)

qihqi committed:
Fix jit can not recognize the imported function

Summary:
Fixes https://github.com/pytorch/pytorch/issues/38964

Test Plan:
Unit test

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

qihqi committed:
Update on "Fix jit can not recognize the imported function"


Summary:
Fixes https://github.com/pytorch/pytorch/issues/38964

Test Plan:
Unit test

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

qihqi commented:
@qihan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30311185).

qihqi committed:
Update on "Fix jit can not recognize the imported function"


Summary:
Fixes https://github.com/pytorch/pytorch/issues/38964

Test Plan:
Unit test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30311185](https://our.internmc.facebook.com/intern/diff/D30311185)

[ghstack-poisoned]

zhxchen17 commented on torch/jit/_script.py:
```diff
@@ -1263,7 +1263,7 @@ def forward(self, input):
             )
         if _rcb is None:
-            _rcb = _jit_internal.createResolutionCallbackFromFrame(_frames_up + 1)
+            _rcb = _jit_internal.createResolutionCallbackFromClosure(obj)
```
I'm surprised that this actually works before, considering the difference between dynamic and lexical resolution. 

Could we rename them to `createResolutionCallbackFromLexicalScope` and `createResolutionCallbackFromDynamicScope` to avoid some misuse in the future?

qihqi replied:
hmmmm not really, createResolutionCallbackFromFrame doesn't give you the lexical scope of the class; but the lexical scope at the point torch.jit.script is called.

zhxchen17 replied:
Update: after an online discussion, we think it's worthwhile to investigate why createResolutionCallbackFromFrame() was there in the first place. If it doesn't make sense we can just remove it, so that the new logic will be much simpler. If it's there for a reason, then I'm not keen on the renaming thing.

qihqi replied:
So turns out createResolutionCallbackFromFrame cannot be removed because it is used to create compiled methods from raw source (as string) as well. In that case, there is no static scope because we dont know where the string come from.

zhxchen17 commented on torch/_jit_internal.py:
```diff
@@ -160,15 +160,36 @@ def __getattr__(self, key):
-
     for index, captured_name in enumerate(fn.__code__.co_freevars):
-        captures[captured_name] = fn.__closure__[index].cell_contents
+        try:
```
when we have ValueError?

qihqi replied:
This happens when torch.jit.script is used as decorator and the a method of class refers the class by name. Added comments.

zhxchen17 replied:
I think it's better to use hasattr() to check existence of cell_contents, that way the intention is more clear and we can avoid accidentally capture unintended errors.

qihqi replied:
hasattr returns True though... i.e. that field exists, it just throws when accessed before it's setup completely :(

zhxchen17 reviewed (commented):
looks great! just some concerns.

qihqi committed:
Update on "Fix jit can not recognize the imported function"


Summary:
Fixes https://github.com/pytorch/pytorch/issues/38964

Test Plan:
Unit test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30311185](https://our.internmc.facebook.com/intern/diff/D30311185)

[ghstack-poisoned]

zhxchen17 reviewed (approved):
I think the fix is reasonable.

qihqi committed:
Update on "Fix jit can not recognize the imported function"


Summary:
Fixes https://github.com/pytorch/pytorch/issues/38964

Test Plan:
Unit test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30311185](https://our.internmc.facebook.com/intern/diff/D30311185)

[ghstack-poisoned]

qihqi commented:
@qihqi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30311185).
